### PR TITLE
is_prime for ideals uses factorization, can be VERY slow

### LINKED ISSUE #33360: is_prime for ideals uses factorization, can be VERY slow
<div id="comment:0"></div>

The implementation of `is_prime()` for ideals uses pari `idealfactor()`. This is a bad idea if the norm of the ideal is hard to factor.

Pari has `ismaximalideal()`, but it is broken in pari 2.13.3. As a workaround, since a prime ideal has prime power norm, we avoid factorization except on this (easy) case.

---

Was: VERY slow doctest in `sage/rings/number_field/number_field.py`

With sage 9.5:

```
$ time sage -t --long --warn-long 60.0 --random-seed=40834229707572602474454926153679777311 src/sage/rings/number_field/number_field.py 
Running doctests with ID 2022-02-16-15-58-06-58d0ee96.
Using --optional=build,pip,sage,sage_spkg,void
Features to be detected: 4ti2,benzene,bliss,buckygen,conway_polynomials,csdp,database_cremona_ellcurve,database_cremona_mini_ellcurve,database_jones_numfield,database_knotinfo,dvipng,graphviz,imagemagick,jupymake,kenzo,latte_int,lrslib,mcqd,meataxe,pandoc,pdf2svg,plantri,pynormaliz,rubiks,sage.combinat,sage.geometry.polyhedron,sage.graphs,sage.plot,sage.rings.number_field,sage.rings.real_double,sage.symbolic,sage_numerical_backends_coin,sagemath_doc_html,sphinx,tdlib
Doctesting 1 file.
sage -t --long --warn-long 60.0 --random-seed=40834229707572602474454926153679777311 src/sage/rings/number_field/number_field.py
**********************************************************************
File "src/sage/rings/number_field/number_field.py", line 2180, in sage.rings.number_field.number_field.NumberField_generic.?
Warning, slow doctest:
    while not K.random_element().is_prime():  # long time
        pass
Test ran for 1319.62 s
    [2265 tests, 1367.95 s]
----------------------------------------------------------------------
All tests passed!
----------------------------------------------------------------------
Total time for all tests: 1368.3 seconds
    cpu time: 1364.7 seconds
    cumulative wall time: 1367.9 seconds
Features detected for doctesting: 

real	22m51.641s
user	22m47.153s
sys	0m1.272s
```

CC:  @yyyyx4 @fchapoton

Component: **number fields**

Author: **Gonzalo Tornaría, Lorenz Panny**

Reviewer: **Lorenz Panny**

_Issue created by migration from https://trac.sagemath.org/ticket/33360_

### PR DESCRIPTION
**Pull request by yyyyx4**
Fixes #33360.

### CODE REVIEW

File: src/sage/rings/number_field/number_field_ideal.py [hunk 1]
```diff
@@ -995,16 +995,36 @@ def is_prime(self):
...
-                self._pari_prime = None
-            else:
-                self._pari_prime = F[0][0]._pari_prime
-            return self._pari_prime is not None
+            pass
+
+        K = self.number_field().pari_nf()
+        I = self.pari_hnf()
+
+        self._pari_prime = K.idealismaximal(I) or None
```
Comment (tornaria) [inline src/sage/rings/number_field/number_field_ideal.py]: Do we have to worry about thread-safety? There's a window of time here were the cache in the ideal thinks it is a prime but it has not been certified yet.

Maybe better to save this to a local variable, then certify the primality, and then fill the object?

Comment (yyyyx4) [reply src/sage/rings/number_field/number_field_ideal.py]: I don't think Sage generally cares about thread safety, does it?

**Comment by yyyyx4** [comment]
The single test failure is unrelated and will be fixed by #35017.

File: src/sage/rings/number_field/number_field_ideal.py (hunk 1 above)
Comment (tornaria) [reply src/sage/rings/number_field/number_field_ideal.py]: > I don't think Sage generally cares about thread safety, does it?

Why not?

Comment (yyyyx4) [reply src/sage/rings/number_field/number_field_ideal.py]: I don't know, but I suspect the reason is that CPython's `threading` implementation does not provide multi-core parallelism.

**Comment by tornaria** [comment]
I'm thinking something simple like this:
```diff
diff --git a/src/sage/rings/number_field/number_field_ideal.py b/src/sage/rings/number_field/number_field_ideal.py
index 1a9d11aec68..8ca1f958039 100644
--- a/src/sage/rings/number_field/number_field_ideal.py
+++ b/src/sage/rings/number_field/number_field_ideal.py
@@ -1017,13 +1017,15 @@ class NumberFieldIdeal(Ideal_generic):
         K = self.number_field().pari_nf()
         I = self.pari_hnf()
 
-        self._pari_prime = K.idealismaximal(I) or None
+        candidate = K.idealismaximal(I) or None
 
         # PARI uses probabilistic primality testing inside idealismaximal().
-        if self._pari_prime \
-                and get_flag(None, 'arithmetic') \
-                and not self._pari_prime[0].isprime():
-            self._pari_prime = None
+        if get_flag(None, 'arithmetic'):
+            # proof required, check using isprime()
+            if candidate and not candidate[0].isprime():
+                candidate = None
+
+        self._pari_prime = candidate
 
         return self._pari_prime is not None
 
```
What do you think? You are probably right that there is no way to trigger anything bad here, but it still smells to me. It also feels easier to read if the condition `get_flag(None, 'arithmetic')` is split and a comment added since that is a pretty opaque name for a function. I couldn't understand what the first parameter is for, and why can't the function be named `require_proof` used as in `require_proof('arithmetic')` or `require_proof_for('arithmetic')` either of which reads more naturally and clearly indicates the meaning.

In any case, do you think it's better to squash everything into a single commit? The final diff is much simpler than 5-6 commits.

**Comment by yyyyx4** [comment]
Okay, all done, including squashing.

**Review by roed314 — APPROVED** [review]
LGTM

**Review by tornaria — APPROVED** [review]

**Pull request merged**